testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java

```diff
@@ -502,6 +502,47 @@ public void testSortItemsReflectedInExplain()
                 expectedPattern);
     }
 
+    // CAST(a_varchar AS date) = DATE '...' is an interesting condition which may want to be optimized by the engine or a connector
+    @Test
+    public void testVarcharCastToDateInPredicate()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE_WITH_DATA));
+
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_varchar_as_date_predicate",
+                "(a varchar)",
+                List.of(
+                        "'999-09-09'",
+                        "'1005-09-09'",
+                        "'2005-06-06'", "'2005-06-6'", "'2005-6-06'", "'2005-6-6'", "' 2005-06-06'", "'2005-06-06 '", "' +2005-06-06'", "'02005-06-06'",
+                        "'2005-09-06'", "'2005-09-6'", "'2005-9-06'", "'2005-9-6'", "' 2005-09-06'", "'2005-09-06 '", "' +2005-09-06'", "'02005-09-06'",
+                        "'2005-09-09'", "'2005-09-9'", "'2005-9-09'", "'2005-9-9'", "' 2005-09-09'", "'2005-09-09 '", "' +2005-09-09'", "'02005-09-09'",
+                        "'2005-09-10'", "'2005-9-10'", "' 2005-09-10'", "'2005-09-10 '", "' +2005-09-10'", "'02005-09-10'",
+                        "'9999-09-09'",
+                        "'99999-09-09'"))) {
+            for (String date : List.of("2005-09-06", "2005-09-09", "2005-09-10")) {
+                for (String operator : List.of("=", "<=", "<", ">", ">=", "!=", "IS DISTINCT FROM", "IS NOT DISTINCT FROM")) {
+                    assertThat(query("SELECT a FROM %s WHERE CAST(a AS date) %s DATE '%s'".formatted(table.getName(), operator, date)))
+                            .hasCorrectResultsRegardlessOfPushdown();
+                }
+            }
+        }
+
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_varchar_as_date_predicate",
+                "(a varchar)",
+                List.of("'2005-06-bad-date'", "'2005-09-10'"))) {
+            assertThatThrownBy(() -> query("SELECT a FROM %s WHERE CAST(a AS date) < DATE '2005-09-10'".formatted(table.getName())))
+                    .hasMessage("Value cannot be cast to date: 2005-06-bad-date");
+            assertThatThrownBy(() -> query("SELECT a FROM %s WHERE CAST(a AS date) = DATE '2005-09-10'".formatted(table.getName())))
+                    .hasMessage("Value cannot be cast to date: 2005-06-bad-date");
```

Comment:
If we want to guarantee this, then I don't think any improvement like #12925 is possible.

@martint @sopel39 @losipiuk @alexjo2144

Reply:
> If we want to guarantee this, then I don't think any improvement like https://github.com/trinodb/trino/issues/12925 is possible.

Why? Because we might miss the error in data?

Reply:
@sopel39 yes, but idk whether it matters. This is a question.

Reply:
@martint do you have an opinion?

Reply:
Indeed, that's hard/impossible to guarantee.

In Trino, a query that should succeed according to standard SQL semantics will succeed and produce correct results. A query that might fail due to bad data or bad assumptions about data may or may not fail depending on optimizations performed by the engine and connectors.

A trivial example is the following:

```
SELECT * FROM (
     SELECT ds, 1/x FROM t
) u
WHERE ds = '2022-08-18'
```

Let's say there's a row with values `('2022-01-01', 0)`. According to standard SQL semantics, such query should fail, as the `1/x` term should be evaluated before the data is filtered by the `ds = '2022-08-18`` clause. In Trino, such a query would generally work due to the predicate being pushed down below the `1/x` (and, possibly, into the connector). This is out of necessity: maintaining strict semantics for failures would make Trino unusable slow -- it would need to scan every row every single time and perform operations in the order established by the original query.

Reply:
@martint that's a good example and pretty compelling.
In your example, one column is used to filter out problematic data from the other column, so query (likely) doesn't fail. This is "out of necessity" (agreed) and also should be easy to explain to users.

My question is about case with one column. Does it change anything for you, @martint ?

Reply:
@martint and I discussed this case. The conclusion is that the even in this particular case the failure isn't guaranteed and optimizations like #12925 are OK to do.

Reply:
Kept the test case, but added a comment:
```
// This failure isn't guaranteed. TODO make test more flexible when need arises.
```